### PR TITLE
Code review PR of #789 - HTTP paths for functions

### PR DESCRIPTION
**Pull request by alexellis**
## Description

Closes #789

Docker image for testing: `alexellis/gateway:http-paths`

### CODE REVIEW

**Comment by alexellis** [comment]
Async Path passing is now tested:

```
faas-cli deploy --name env --fprocess env --network func_functions --image functions/alpine:latest

curl -d "" http://127.0.0.1:8080/async-function/env/path_here

func_queue-worker.1.qk96166nz9ji@nuc    | [#3] Received on [faas-request]: 'sequence:6 subject:"faas-request" data:"{\"Header\":{\"Accept\":[\"*/*\"],\"Content-Length\":[\"0\"],\"Content-Type\":[\"application/x-www-form-urlencoded\"],\"User-Agent\":[\"curl/7.55.1\"],\"X-Call-Id\":[\"09577dcb-46bd-473b-9bf6-ede3f44e96a4\"],\"X-Start-Time\":[\"1535102101843119850\"]},\"Host\":\"127.0.0.1:8080\",\"Body\":\"\",\"Method\":\"POST\",\"Path\":\"/path_here\",\"QueryString\":\"\",\"Function\":\"env\",\"CallbackUrl\":null}" timestamp:1535102101843442385 '
```

`"Path\":\"/path_here\"`

Also tested with `?qs=1` and saw that passed through.

Path is empty when nothing is passed after `/async-function/env`

I tested the path work via Ivana's Ruby guestbook including GET/POST endpoints at different paths.

```
faas-cli deploy --name guestbook --network func_functions --image docwareiy/guestbook:0.1
```

screen shot 2018-08-24 at 10 17 16 am

**Comment by alexellis** [comment]
@telackey please confirm that the original intent is still matched in this follow-up PR.

* I tested a guestbook microservice in Ruby using GET/POST - worked fine.
* I tested the UI + CLI + paths were passed to the back-end and routed as expected
* I tested the Java of-watchdog template which still worked fine when no path was passed
* I tested async functions include the path variable.

**Comment by ivanayov** [comment]
I tested with `docwareiy/guestbook:0.2` for path parameters + url query:
http://127.0.0.1:8080/function/guestbook/signatures?guest=foo%20bar&firstname=foo&lastname=bar
Output:
`Visitor foo bar has signed the Guestbook.`.

http://127.0.0.1:8080/function/guestbook/signatures?guest=test%20bar&firstname=test&lastname=bar
Output:
`Visitor test bar has not signed the Guestbook.`.

Also tested async functions with path parameters.

**Comment by ivanayov** [comment]
Using this gateway, function invocation count does not increase. Tested both in the UI and Prometheus by `gateway_function_invocation_total{function_name="figlet"}`.

After reverting to `openfaas/gateway:0.8.11` it works as expected.

File: gateway/handlers/forwarding_proxy.go [hunk 1]
```diff
@@ -25,12 +37,17 @@ type BaseURLResolver interface {
...
+}
+
 // MakeForwardingProxyHandler create a handler which forwards HTTP requests
-func MakeForwardingProxyHandler(proxy *types.HTTPClientReverseProxy, notifiers []HTTPNotifier, baseURLResolver BaseURLResolver) http.HandlerFunc {
+func MakeForwardingProxyHandler(proxy *types.HTTPClientReverseProxy, notifiers []HTTPNotifier, baseURLResolver BaseURLResolver, urlPathTransformer URLPathTransformer) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		baseURL := baseURLResolver.Resolve(r)
 
-		requestURL := r.URL.Path
+		requestURL := urlPathTransformer.Transform(r)
```
Comment (ivanayov) [inline gateway/handlers/forwarding_proxy.go]: This transform method causes the bug with invocation count.

Comment (ivanayov) [reply gateway/handlers/forwarding_proxy.go]: Trying to fix

**Comment by ivanayov** [comment]
Derek add label: blocked

**Comment by ivanayov** [comment]
Derek remove label: blocked

**Comment by telackey** [comment]
@alexellis 

>please confirm that the original intent is still matched in this follow-up PR.

Yes, looks good.

**Comment by alexellis** [comment]
Appreciated Lucas.

**Comment by alexellis** [comment]
@telackey do the upstream templates have the ability to pass on a path to their handlers yet?

**Pull request merged**